src/main/kotlin/org/veupathdb/lib/rabbit/jobs/pools/exec/JobQueueExecutorPool.kt

```diff
@@ -0,0 +1,232 @@
+ * val executor = JobQueueExecutorPool(config)
+ *
+ * // Start the executor pool's workers
+ * executor.start()
+ *
+ * // Gracefully stop the executor pool
+ * executor.stop()
+ *
+ * // Immediately stop the executor pool (interrupting any running jobs)
+ * executor.stopNow()
+ * ```
+ *
+ * @author Elizabeth Paige Harper [foxcapades.io@gmail.com]
+ * @since 2.0.0
+ */
+internal class JobQueueExecutorPool(private val config: ExecutorPoolConfig) {
+  private val logger = LoggerFactory.getLogger(javaClass)
+
+  private val closing = AtomicBoolean(false)
+  private val name = "${config.queueName}-consumer-pool"
+  private val threads = Executors.newFixedThreadPool(config.poolSize)
+
+  private var head: Executor? = null
+  private var tail: Executor? = null
+  private var counter = 0u
+
+  init {
+    if (config.poolSize !in 1..64)
+      throw IllegalArgumentException("invalid size value ${config.poolSize}, must be in the range [1, 64]")
+  }
+
+  /**
+   * Starts the executor pool's workers.
+   *
+   * Until this method is called, the worker pool will not receive or react to
+   * any job queue messages.
+   */
+  fun start() {
+    for (i in 0 ..< config.poolSize)
+      add(createNew().also { it.init(config.queueName, onConsumerShutdown(it)) })
+  }
+
+  /**
+   * Attempts to gracefully shut down the executor pool.
+   *
+   * If [block] is `true`, this method will block the current thread until the
+   * [killAfter] duration has been reached, at which point the executor pool
+   * will request a force-shutdown of the underlying thread pool, aborting any
+   * jobs still in progress at that point.  If [blockAfterKill] is also `true`
+   * at this point, the method will continue to block after the force-shutdown
+   * request until the last job has ended.
+   *
+   * If [block] is `false`, this method will not block.  It will request the
+   * underlying thread pool shut down, then return immediately.  In this case
+   * the values of [killAfter] and [blockAfterKill] are ignored.
+   *
+   * @param block Whether this method should block the current thread until the
+   * underlying thread pool has gracefully shut down.
+   *
+   * @param killAfter Max duration to wait for a graceful shutdown before
+   * attempting to abort remaining in-progress job executions.
+   *
+   * @param blockAfterKill Whether this method should continue to block after
+   * the [killAfter] duration has passed.
+   */
+  fun stop(
+    block: Boolean = false,
+    killAfter: Duration = 10_000.days,
+    blockAfterKill: Boolean = true,
+  ): StopCode {
+    stopExecutors()
+    threads.shutdown()
+
+    if (!block)
+      return if (threads.isTerminated) StopCode.Graceful else StopCode.Unknown
+
+    if (threads.awaitTermination(killAfter.inWholeMilliseconds, TimeUnit.MILLISECONDS))
+      return StopCode.Graceful
+
+    threads.shutdownNow()
+    silently(config.shutdownCB)
+    if (blockAfterKill)
+      threads.awaitTermination(243256, TimeUnit.DAYS)
+
```

Comment:
Is there any significance to this wait time, or is just effectively "wait indefinitely" and there's no other way to do that?

Reply:
It is just wait indefinitely.  I forget what that specific number 'means', but that meaning was a joke anyway.  I will add a comment explaining it.